### PR TITLE
fix: add conditional to database different postgres

### PR DESCRIPTION
**Pull request by analuramalho**
Signed-off-by: Ana Luiza Ramalho dos Santos <analuizaramalho9@gmail.com>

## Hey, I just made a Pull Request!

When the database is MySQL the ".returning()" on the end of the query didn't work to this type so I add a conditional to fix it. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
@freben gonna punt this to you! :pray:

File: .changeset/rare-buses-swim.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/backend-tasks': minor
```
Comment (freben) [inline .changeset/rare-buses-swim.md:2]: This should be minor for non-breaking changes

File: packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.ts [hunk 2]
```diff
@@ -53,15 +53,31 @@ export class PluginTaskSchedulerJanitor {
 
   private async runOnce() {
     const dbNull = this.knex.raw('null');
+    let tasks = [];
```
Comment (freben) [inline packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.ts:56]: Can you make sure to put an explicit type on this just to make sure that both branches return the right thing and that readers know upfront what's going in here? Like

```ts
let tasks: Array<{ id: string }>;
```

should be possible to not assign `[]` to it too, I think, the compiler should make sure that all branches assign to it properly before reading.

File: packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.ts [hunk 3]
```diff
@@ -53,15 +53,31 @@ export class PluginTaskSchedulerJanitor {
...
 
-    const tasks = await this.knex<DbTasksRow>(DB_TASKS_TABLE)
-      .where('current_run_expires_at', '<', this.knex.fn.now())
-      .update({
-        current_run_ticket: dbNull,
-        current_run_started_at: dbNull,
-        current_run_expires_at: dbNull,
-      })
-      .returning(['id']);
+    if (configClient.includes('sqlite3') || configClient.includes('mysql')) {
```
Comment (freben) [inline packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.ts:59]: Would you mind adding a test file for this class? You can take inspiration from the other test files nearby such as `TaskWorker.test.ts` where it creates some test databases to run against. If you have the `MYSQL_8` constant in the list, you know that the test is run against it and exercises this exact code

**Review by freben — COMMENTED** [review]
Thanks!

File: .changeset/rare-buses-swim.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/backend-tasks': major
```
Comment (freben) [inline .changeset/rare-buses-swim.md:2]: Not major though! Just minor. Also remember the DCO signature whenever you make commits.

File: .changeset/rare-buses-swim.md (hunk 1 above)
Comment (freben) [reply .changeset/rare-buses-swim.md:2]: Sorry, I meant patch - since this is on a 0.x version still

**Review by freben — COMMENTED** [review]
Hey, just a gentle ping here

**Comment by freben** [comment]
reminder to preferably put a test in place too

**Review by freben — APPROVED** [review]
Alright, made a fix and waiting for it to go green

File: packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.test.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+const getTask = async (knex: Knex): Promise<DbTasksRow> => {
+  return (await knex<DbTasksRow>(DB_TASKS_TABLE))[0];
+};
+
+describe('PluginTaskSchedulerJanitor', () => {
+  const logger = getVoidLogger();
+  const databases = TestDatabases.create({
+    ids: [
+      /* 'MYSQL_8' not supported yet */
```
Comment (analuramalho) [inline packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.test.ts:41]: the plugins/catalog-backend/src/database/DefaultProcessingDatabase.test.ts uses the MYSQL_8.
why in this case the mysql 8 not work ?

Comment (freben) [reply packages/backend-tasks/src/tasks/PluginTaskSchedulerJanitor.test.ts:41]: That's the catalog, not the backend-tasks package. Every plugin that has any database commands, needs to separately make sure that their stuff works with the required database engines. Mysql is very optional, and almost nothing works with it; postgres is the only strictly supported engine in production. Feel free to try the tests with that commented-out ID added to see what mysql doesn't like about the migrations in their current state.

**Pull request merged**